File: DSA/LakshaKundra_590013237/Day_50_Question_2.py
from collections import Counter

def count_anagrams(text, pattern):
    k = len(pattern)
    pattern_count = Counter(pattern)
    window_count = Counter(text[:k])
    count = 0
    if window_count == pattern_count:
        count += 1
    for i in range(k, len(text)):
        window_count[text[i]] += 1
        window_count[text[i - k]] -= 1
        if window_count[text[i - k]] == 0:
            del window_count[text[i - k]]
        if window_count == pattern_count:
            count += 1
    return count

print(count_anagrams("forxxorfxdofr", "for"))  # Output: 3
print(count_anagrams("aabaabaa", "aaba"))      # Output: 4

File: DSA/Gautam_Shyamnani_590013078/Day50Question2.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

int countAnagrams(string text, string pattern) {
    int count = 0;
    int k = pattern.length();
    unordered_map<char, int> patMap, winMap;

    // Frequency map for pattern
    for (char ch : pattern) patMap[ch]++;

    int match = 0;
    for (int i = 0; i < text.length(); i++) {
        // Add current character to window map
        winMap[text[i]]++;

        // Remove character that's now outside the window
        if (i >= k) {
            if (winMap[text[i - k]] == 1)
                winMap.erase(text[i - k]);
            else
                winMap[text[i - k]]--;
        }

        // Compare maps when window size equals pattern length
        if (i >= k - 1 && winMap == patMap)
            count++;
    }

    return count;
}

int main() {
    string text1 = "forxxorfxdofr", pattern1 = "for";
    string text2 = "aabaabaa", pattern2 = "aaba";

    cout << "Anagram count (text1): " << countAnagrams(text1, pattern1) << endl; // Output: 3
    cout << "Anagram count (text2): " << countAnagrams(text2, pattern2) << endl; // Output: 4

    return 0;
}
